kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+    err = ("dns %s error: %s %s"):format(
+            answers.errcode < CACHE_ONLY_ERROR_CODE and "server" or "client",
+            answers.errcode, answers.errstr)
+    table_insert(tries, { name .. ":" .. TYPE_TO_NAME[qtype], err })
+  end
+
+  return answers
+end
+
+
+-- resolve all `name`s and return first usable answers
+local function resolve_query_names(self, names, qtype, tries)
+  local answers, err
+
+  for _, qname in ipairs(names) do
+    answers, err = resolve_query(self, qname, qtype, tries)
+
+    -- severe error occurred
+    if not answers then
+      return nil, err
+    end
+
+    if not answers.errcode then
+      return answers, nil, answers.ttl
+    end
+  end
+
+  -- not found in the search iteration
+  return answers, nil, answers.ttl
+end
+
+
+local function resolve_query_types(self, name, qtype, tries)
+  local names = search_names(name, self.resolv, self.hosts)
+  local answers, err, ttl
+
+  -- the specific type
+  if qtype ~= TYPE_A_AAAA then
+    return resolve_query_names(self, names, qtype, tries)
+  end
+
+  -- query A or AAAA
+  if self.enable_ipv4 then
+    answers, err, ttl = resolve_query_names(self, names, TYPE_A, tries)
+    if not answers or not answers.errcode then
+      return answers, err, ttl
+    end
+  end
+
+  if self.enable_ipv6 then
+    answers, err, ttl = resolve_query_names(self, names, TYPE_AAAA, tries)
+  end
+
+  return answers, err, ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype)
+  if premature then
+    return
+  end
+
+  local tries = setmetatable({}, _TRIES_MT)
+  local answers = resolve_query_types(self, name, qtype, tries)
+  if not answers or answers.errcode then
+    log(DEBUG, PREFIX, "failed to update stale DNS records: ", tostring(tries))
+    return
+  end
+
+  log(DEBUG, PREFIX, "update stale DNS records: ", #answers)
+  self.cache:set(key, { ttl = answers.ttl }, answers)
+end
+
+
+local function start_stale_update_task(self, key, name, qtype)
+  stats_increment(self.stats, key, "stale")
+
+  local ok, err = timer_at(0, stale_update_task, self, key, name, qtype)
+  if not ok then
+    log(ALERT, PREFIX, "failed to start a timer to update stale DNS records: ", err)
+  end
+end
+
+
+local function check_and_get_ip_answers(name)
+  -- TODO: use is_valid_ipv4 from kong/tools/ip.lua instead
+  if name:match("^%d+%.%d+%.%d+%.%d+$") then  -- IPv4
+    return {
+      { name = name, class = 1, type = TYPE_A, address = name },
+    }
+  end
+
+  if name:find(":", 1, true) then             -- IPv6
+    return {
+      { name = name, class = 1, type = TYPE_AAAA, address = ipv6_bracket(name) },
+    }
+  end
+
+  return nil
+end
+
+
+local function resolve_callback(self, name, qtype, cache_only, tries)
+  -- check if name is ip address
+  local answers = check_and_get_ip_answers(name)
+  if answers then -- domain name is IP literal
+    answers.ttl = LONG_LASTING_TTL
+    answers.expire = now() + answers.ttl
+    return answers, nil, tries
```

Comment:
Should be `return answers, nil, answers.ttl`?

Reply:
```suggestion
    return answers, nil, answers.ttl
```

Reply:
done

Reply:
See this line https://github.com/Kong/kong/blob/master/kong/resty/mlcache/init.lua#L694, mlcache will ignore non-number type ttl value, so this bug has not been catched by our test cases.

Good catch, QI!